Add documentation to GodotSharp, Fix PackedInt64Array typo

raulsntos opened the pull request:
Follow-up to #40218, #50867 and #51044

- Adds documentation to almost every class and member in `GodotSharp` (Related to #39703 but I'm not sure if this is enough to close it).
- Fixes a typo in `bindings_generator.h`
- Fixes the documentation generation for `[constant]` tags in `bindings_generator.cpp`

raulsntos committed:
Fix PackedInt64Array typo in bindings_generator.h

aaronfranke reviewed (commented):
Excellent work!

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/GD.cs:
It seems quite a few methods are missing `<returns>`, sometimes it might not be necessary to have one since the `<summary>` is clear enough but I can make sure there's always a `<returns>` if that's preferred for consistency.

aaronfranke replied:
I've been including it everywhere even if not necessary to be consistent with the auto-generated XML blocks when typing `///`. I think we should be consistent and have it everywhere.

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Color.cs:
```diff
@@ -987,32 +1002,49 @@ public override bool Equals(object obj)
         }
 
+        /// <summary>
+        /// Returns <see langword="true"/> if this color and <paramref name="other"/> are equal
+        /// </summary>
+        /// <param name="other">The other <see cref="Color"/> to compare.</param>
+        /// <returns>Whether or not the <see cref="Color"/>s are equal.</returns>
```
I'm fine with either, my personal preference would be to use it whenever possible since an IDE would be able to rename it everywhere when refactoring as well as link to it, but I guess it's unnecessary in this case.

neikeq replied:
That being said, the benefits of these docs outweigh this, so it doesn't prevent merging the PR.

aaronfranke replied:
It seems in the .NET docs that the types are indeed used to complement other nouns rather than being nouns by themselves. This does seem a bit strange though. I think I would prefer `this <see cref="Color"/>` to `this color` in the summary even if `this <see cref="Color"/> instance` or similar is better.

aaronfranke replied:
Here's a case where they do use the type as a noun: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.count?view=net-5.0

raulsntos committed:
Fix constant tag documentation in C# bindings generator

aaronfranke commented on modules/mono/glue/GodotSharp/GodotSharp/Core/GD.cs:
This statement isn't quite clear enough on its own, since volume can also mean 3D space. It's likely fine because of the context, but I think we should make it clearer:

```suggestion
        /// <returns>Audio volume as linear energy.</returns>
```

aaronfranke commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -387,7 +387,7 @@ String BindingsGenerator::bbcode_to_xml(const String &p_bbcode, const TypeInterf
 					xml_output.append("</c>");
 				}
-			} else if (link_tag == "const") {
+			} else if (link_tag == "constant") {
```
@neikeq

neikeq replied:
I just checked and there's no `[const ` occurrences. It's `constant` indeed.

aaronfranke reviewed (approved):
What is the reason that this PR is marked as draft? It looks good to me aside from the discussion above with Color. I approve of the changes in the `.cs` files, and the typo fix looks good. I'd like @neikeq to look at this, specifically the below:

raulsntos commented:
> What is the reason that this PR is marked as draft?

Since this PR changes so many files and #50872 will probably conflict I thought you may want to merge that one first and then I can rebase this one.

neikeq reviewed (approved):
This is huge, thanks!

raulsntos committed:
Add documentation to GodotSharp

- Adds documentation to almost every class and member in `GodotSharp`
- Fixes some old documentation to more closely follow the XML comments convention

neikeq commented:
#50872 is now merged, so this PR is good to merge as well once rebased.

neikeq commented:
Thanks!
